### PR TITLE
Feature/mob 4160 extract correlation - APPROACH A

### PR DESCRIPTION
**Pull request by BartlomiejSopalaKarhoo**
## JIRA
[MOB-4166](https://karhoo.atlassian.net/browse/MOB-4166)

## DESCRIPTION

APPROACH A

This pr presents network sdk will change if we decide to extend current Result structure and create tuples in success/failure case, and add correlationId as a second component.

related [UI SDK pull request](https://github.com/karhoo/karhoo-ios-ui-sdk/pull/350)

## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix 
- [ ] New feature 
- [ ] Tech Debt
- [ ] Breaking change 
- [ ] Documentation Update
- [ ] Other: *define what type of change it is*

## CHECKLIST

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging code._
Please review the guidelines for contributing to this repository.

- [ ] All work relating to the ticket is complete (Code complete/Acceptance Criteria met)
- [ ] Lint and unit tests pass locally (if appropriate)
- [ ] Added tests that prove the fix is effective or that feature works (if appropriate)
- [ ] Documentation (if appropriate)

## SCREENSHOTS
[Add screenshots if any for changes you have made to the App/UI]

### CODE REVIEW

File: KarhooSDK/Api/Util/Result.swift [hunk 1]
```diff
@@ -41,13 +41,22 @@ public enum Result<T> {
             return false
         }
     }
+    
+    public func correlationId() -> String? {
```
[the remark this answers is not in the document]
Comment (BartlomiejSopalaKarhoo) [reply KarhooSDK/Api/Util/Result.swift:45]: Yes, I know and I agree, but in this enum are other methods like successValue() and errorValue() without "get" prefix, that's why I decided to not use get also for this one.

File: KarhooSDK/Service/Analytics/AnalyticsConstants.swift [hunk 2]
```diff
@@ -69,19 +69,32 @@ public enum AnalyticsConstants {
...
         case ridesPastTripsOpened = "RIDES_PAST_LIST"
         case ridesUpcomingTripsOpened = "RIDES_UPCOMING_LIST"
         case ridesUpcomingTrackTripClicked = "RIDES_UPCOMING_VEHICLE_TRACKING_CLICKED"
         case ridesUpcomingContactFleetClicked = "RIDES_UPCOMING_CONTACT_FLEET_CLICKED"
         case ridesUpcomingContactDriverClicked = "RIDES_UPCOMING_CONTACT_DRIVER_CLICKED"
         case trackingContactDriverClicked = "VEHICLE_TRACKING_SCREEN_CONTACT_DRIVER_CLICKED"
-        
+
+        // PAYMENT STATUS
+        case paymentSucceed = "BOOKING_PAYMENT_SUCCESS"
```
Comment (dianakarh) [inline KarhooSDK/Service/Analytics/AnalyticsConstants.swift:81]: Olivier updated these names to `success`, `failure`, and `requested`

File: KarhooSDK/Service/Payment/AdyenPaymentMethods/KarhooAdyenPaymentMethodsInteractor.swift [hunk 3]
```diff
@@ -46,8 +46,8 @@ final class KarhooAdyenPaymentMethodsInteractor: AdyenPaymentMethodsInteractor {
 
     private func handle(_ response: Result<HttpResponse>, interactorCallback: CallbackClosure<DecodableData>) {
         switch response {
-        case .failure(let error): interactorCallback(.failure(error: error))
-        case .success(let response): interactorCallback(.success(result: DecodableData(data: response.data)))
+        case .failure(let error): interactorCallback(.failure(error: error.error))
+        case .success(let result): interactorCallback(.success(result: DecodableData(data: result.result.data)))
```
Comment (dianakarh) [inline KarhooSDK/Service/Payment/AdyenPaymentMethods/KarhooAdyenPaymentMethodsInteractor.swift:50]: don't really like this `result.result` bit. also the warnings need to be sorted out. what if we go with this?
```
case .failure(let error, _): interactorCallback(.failure(error: error))
case .success(let result, _): interactorCallback(.success(result: DecodableData(data: result.data)))
```

File: KarhooSDK/Api/Util/Result.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (dianakarh) [reply KarhooSDK/Api/Util/Result.swift:45]: Let's go like this: let's mark the existing `successValue()`, `errorValue()` (etc if there are pre-existing others) as deprecated and create the proper ones as well. In the deprecated warning mention that these will be made obsolete in a future release and that they should use xxx instead.

File: KarhooSDK/Service/Payment/AdyenPaymentMethods/KarhooAdyenPaymentMethodsInteractor.swift (hunk 3 above)
Comment (BartlomiejSopalaKarhoo) [reply KarhooSDK/Service/Payment/AdyenPaymentMethods/KarhooAdyenPaymentMethodsInteractor.swift:50]: then we will wet warning that accesing tuplets in that way is deprecated :/

File: KarhooSDK/Api/Util/Result.swift [hunk 4]
```diff
@@ -9,23 +9,33 @@
 import Foundation
 
 public enum Result<T> {
-    case success(result: T)
-    case failure(error: KarhooError?)
+    case success(result: T, correlationId: String? = nil)
+    case failure(error: KarhooError?, correlationId: String? = nil)
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooSDK/Api/Util/Result.swift:13]: I'd like to empathize how much I dislike the need to do such hacks. We need to do it, but I'd like to make it clear :D

File: KarhooSDK/Service/Payment/AdyenPaymentMethods/KarhooAdyenPaymentMethodsInteractor.swift (hunk 3 above)
Comment (AleksanderWedrychowskiKarhoo) [reply KarhooSDK/Service/Payment/AdyenPaymentMethods/KarhooAdyenPaymentMethodsInteractor.swift:50]: We could rename the companion value identifier of success case

**Review by AleksanderWedrychowskiKarhoo — DISMISSED** [review]

File: KarhooSDK/Service/Analytics/AnalyticsConstants.swift [hunk 5]
```diff
@@ -37,8 +37,8 @@ public enum AnalyticsConstants {
         case pickupReverseGeocodeRequested = "pickup_reverse_geocode_requested"
         case pickupReverseGeocodeResponded = "pickup_reverse_geocode_responded"
         case preebookTimeSet = "preebook_time_set"
-        case userCardRegistered = "user_card_registered"
-        case userCardRegistrationFailed = "user_card_registration_failed"
+        case userCardRegistered = "user_card_registered" // ??
+        case userCardRegistrationFailed = "user_card_registration_failed" // ??
```
Comment (dianakarh) [inline KarhooSDK/Service/Analytics/AnalyticsConstants.swift:41]: is the `//??` leftover cleanup?

File: KarhooSDK/Service/Analytics/AnalyticsConstants.swift [hunk 6]
```diff
@@ -69,19 +69,32 @@ public enum AnalyticsConstants {
...
         case ridesPastTripsOpened = "RIDES_PAST_LIST"
         case ridesUpcomingTripsOpened = "RIDES_UPCOMING_LIST"
         case ridesUpcomingTrackTripClicked = "RIDES_UPCOMING_VEHICLE_TRACKING_CLICKED"
         case ridesUpcomingContactFleetClicked = "RIDES_UPCOMING_CONTACT_FLEET_CLICKED"
         case ridesUpcomingContactDriverClicked = "RIDES_UPCOMING_CONTACT_DRIVER_CLICKED"
         case trackingContactDriverClicked = "VEHICLE_TRACKING_SCREEN_CONTACT_DRIVER_CLICKED"
-        
+
+        // PAYMENT STATUS
+        case bookingSuccess = "BOOKING_SUCCESS"
```
Comment (dianakarh) [inline KarhooSDK/Service/Analytics/AnalyticsConstants.swift:81]: some event names are lower case, others are uppercase. we need a cleanup for this @olivierkarhoo. same situation on android.

File: KarhooSDK/Service/Analytics/AnalyticsConstants.swift (hunk 5 above)
Comment (BartlomiejSopalaKarhoo) [reply KarhooSDK/Service/Analytics/AnalyticsConstants.swift]: my bad, removed

File: KarhooSDK/Service/Analytics/AnalyticsConstants.swift (hunk 6 above)
Comment (BartlomiejSopalaKarhoo) [reply KarhooSDK/Service/Analytics/AnalyticsConstants.swift:81]: As we agreed on DSM: all new events with uppercase

**Review by AleksanderWedrychowskiKarhoo — APPROVED** [review]

**Review by dianakarh — APPROVED** [review]

**Pull request merged**